### PR TITLE
Add Segformer

### PR DESCRIPTION
**Pull request by innat**
fix: https://github.com/innat/medic-ai/issues/12

### CODE REVIEW

File: medicai/models/segformer/segformer_model.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        return x
+
+    def resize_to_target(self, x, target_spatial_shape, spatial_dims):
+        current_shape = ops.shape(x)[1:-1]
+        if spatial_dims == 3:
+            size_factors = [
+                int(ops.round(target_spatial_shape[i] / current_shape[i]))
+                for i in range(spatial_dims)
+            ]
+            x = layers.UpSampling3D(size=size_factors)(x)
```
[the remark this answers is not in the document]
Comment (innat) [reply medicai/models/segformer/segformer_model.py:119]: /gemini

Do you mean, we can use `tf.image.resize` on height and width. And later manually resize depth axis with some sort of interpolation method, i.e. 'linear'. 

```python
new_depth, new_height, new_width = 60, 60, 60
input = (20, 40, 40, 3) # depth, height, width, channel
resize_hw = tf.image.resize(input, (new_height, new_width))
resize_dhw = depth_interpolation(resize_hw, new_depth, depth_axis=0)
resize_dhw.shape # (60, 60, 60, 3)
```

File: medicai/models/segformer/segformer_model.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+        return apply
+
+    def linear_embedding(self, x, hidden_dims):
+        spatial_shape_tensor = ops.shape(x)[1:-1]
+        num_patches = int(np.prod(spatial_shape_tensor))
+        x = layers.Reshape((num_patches, ops.shape(x)[-1]))(x)
+        x = layers.Dense(hidden_dims)(x)
+        x = layers.LayerNormalization()(x)
+        return x
```
[the remark this answers is not in the document]
Comment (innat) [reply medicai/models/segformer/segformer_model.py:194]: @gemini-code-assist 

Segformer model is not implemented with Model-subclass API. It's functional model built within class approach. So, Dense and LayerNorm will not create new node every time they're called.

File: medicai/utils/general.py [hunk 3]
```diff
@@ -182,3 +180,56 @@ def _crop_output(
...
+
+        dx = ops.reshape(dx, (1, 1, 1, width, 1))
+        out = (1.0 - dx) * lower_w + dx * upper_w  # (B, out_d, out_h, out_w, C)
+        out = ops.cast(out, original_dtype)
+        return out
+
+    if method == "trilinear":
+        return trilinear_resize(volumes, depth, height, width)
+    else:
+        raise ValueError(f"Unsupported resize method: {method}")
```
[the remark this answers is not in the document]
Comment (innat) [reply medicai/utils/general.py]: /gemini

Recheck your fact. 

```python
keras.ops.image.resize(
    images,
    size,
    interpolation="bilinear",
    antialias=False,
    crop_to_aspect_ratio=False,
    pad_to_aspect_ratio=False,
    fill_mode="constant",
    fill_value=0.0,
    data_format=None,
)

Resize images to size using the specified interpolation method.

Arguments

images: Input image or batch of images. Must be 3D or 4D.
size: Size of output image in (height, width) format.
interpolation: Interpolation method. Available methods are "nearest", "bilinear", and "bicubic". Defaults to "bilinear".
antialias: Whether to use an antialiasing filter when downsampling an image. Defaults to False.
crop_to_aspect_ratio: If True, resize the images without aspect ratio distortion. When the original aspect ratio differs from the target aspect ratio, the output image will be cropped so as to return the largest possible window in the image (of size (height, width)) that matches the target aspect ratio. By default (crop_to_aspect_ratio=False), aspect ratio may not be preserved.
pad_to_aspect_ratio: If True, pad the images without aspect ratio distortion. When the original aspect ratio differs from the target aspect ratio, the output image will be evenly padded on the short side.
fill_mode: When using pad_to_aspect_ratio=True, padded areas are filled according to the given mode. Only "constant" is supported at this time (fill with constant value, equal to fill_value).
fill_value: Float. Padding value to use when pad_to_aspect_ratio=True.
data_format: A string specifying the data format of the input tensor. It can be either "channels_last" or "channels_first". "channels_last" corresponds to inputs with shape (batch, height, width, channels), while "channels_first" corresponds to inputs with shape (batch, channels, height, width). If not specified, the value will default to [keras.config.image_data_format](https://keras.io/api/utils/config_utils#imagedataformat-function).
```

Official link: https://keras.io/api/ops/image/

File: medicai/models/segformer/segformer_layers.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+            # Layer Normalization
+            x = get_norm_layer(norm_name="layer", epsilon=layer_norm_epsilon)(x)
+
+            # Reshape output to a spatial feature map for the next stage.
+            n_patches = ops.shape(x)[1]
+            current_spatial_dims = int(ops.round(n_patches ** (1 / spatial_dims)))
+            current_spatial_dims = [current_spatial_dims] * spatial_dims
+            x = keras.layers.Reshape(
+                current_spatial_dims + [project_dim[i]], name=f"mixvit_features{i+1}"
+            )(x)
```
[the remark this answers is not in the document]
Comment (innat) [reply medicai/models/segformer/segformer_layers.py:512]: /gemini

One of the current assumption of Segfomer arch is that, the input is perfect square or cubic. https://github.com/OSUPCVLab/SegFormer3D/issues/27

To support non-cubick or arbitrary input shape, we didn't explore this yet. Maybe later if needed.

File: medicai/models/segformer/segformer_model.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            x = layers.Dropout(dropout)(x)
+
+            # Final prediction
+            x = get_conv_layer(
+                spatial_dims=spatial_dims, layer_type="conv", filters=num_classes, kernel_size=1
+            )(x)
+            x = get_reshaping_layer(spatial_dims=spatial_dims, layer_type="upsampling", size=4)(x)
+            return x
+
+        return apply
```
[the remark this answers is not in the document]
Comment (innat) [reply medicai/models/segformer/segformer_model.py:185]: /gemini

We've discussed it earlier. To understand you've uderstood, can you implement the SegFormer class fully. Assume other (internal component) already exist.

File: medicai/models/segformer/segformer_layers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (innat) [reply medicai/models/segformer/segformer_layers.py:512]: /gemini

Good catch if set a warning. Maybe we can do this as soon as we get `spatial_dims = len(input_shape) - 1` in segformer class. We can set a value check.

Also, to support arbitrary input shape or (non-square, non-cubic) what are your take (idea) to support here?

File: medicai/models/segformer/segformer_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (innat) [reply medicai/models/segformer/segformer_model.py:202]: @gemini-code-assist

[the remark this answers is not in the document]
Comment (innat) [reply medicai/models/segformer/segformer_model.py:202]: /gemini

As I thought, you wrongly assumed we are building SegFormer with model subclass API, but we are building model with functional api. For that, it is completely fine with current code. Though some may still suggest pre-define the layers at earlier but this will only make code implicit to read. So, we are good at current one.

**Comment by innat** [comment]
/gemini review

Focus on the failthfullness of keras implementation against Segformer3D model https://arxiv.org/pdf/2404.10156
